### ISSUE TITLE
Can't get `systemd-resolved` to use specified DNS servers

### ISSUE THREAD

**Issue by roy-work** [systemd/systemd#31674]
### systemd version the issue has been seen with

247

### Used distribution

minikube

### Linux kernel version used

5.10.57

### CPU architectures issue was seen on

x86_64

### Component

networkctl, resolvectl, systemd-networkd, systemd-resolved

### Expected behaviour you didn't see

My `/etc/resolv.conf` appears thusly:

```
$ ls -l /etc/resolv.conf
lrwxrwxrwx 1 root root 34 Jan 27  2023 /etc/resolv.conf -> ../run/systemd/resolve/resolv.conf
```
Thus I understand it to be managed by `systemd-resolved`. The contents list a DNS server I think is the one minikube supplies via DHCP.

<details>
<summary>The contents of <tt>/etc/resolv.conf</tt></summary>

```
$ cat /etc/resolv.conf
# This is /run/systemd/resolve/resolv.conf managed by man:systemd-resolved(8).
# Do not edit.
#
# This file might be symlinked as /etc/resolv.conf. If you're looking at
# /etc/resolv.conf and seeing this text, you have followed the symlink.
#
# This is a dynamic resolv.conf file for connecting local clients directly to
# all known uplink DNS servers. This file lists all configured search domains.
#
# Third party programs should typically not access this file directly, but only
# through the symlink at /etc/resolv.conf. To manage man:resolv.conf(5) in a
# different way, replace this symlink by a static file or a different symlink.
#
# See man:systemd-resolved.service(8) for details about the supported modes of
# operation for /etc/resolv.conf.

nameserver 192.168.64.1
search .
```

</details>

I think that file *should* contain `nameserver 1.1.1.1`, but it doesn't.

In `/etc/systemd/resolved.conf`, I have the following configuration:

```
[Resolve]
DNS=1.1.1.1
```

Now, AIUI, any resolvers from DHCP _trump_ that one. Thus, I've set `/etc/systemd/network/20-dhcp.network` to,

```
[Match]
Name=eth*

[Network]
DHCP=ipv4
DNS=
LinkLocalAddressing=no

[DHCPv4]
ClientIdentifier=mac
UseDNS=false
```

This is a combination of the default (in minikube), and my modifications. Specifically, I'm setting `UseDNS=false` and `DNS=` (to the empty string), which I thought is the required configuration to get the interface to not use DNS servers received over DNS.

If I run `networkctl status eth0`, I get:
```
● 2: eth0
                     Link File: n/a
                  Network File: /etc/systemd/network/20-dhcp.network
                          Type: ether
                         State: routable (configured)
                    HW Address: 06:6a:7b:c5:3e:76
                           MTU: 1500 (min: 68, max: 65535)
                         QDisc: pfifo_fast
  IPv6 Address Generation Mode: none
          Queue Length (Tx/Rx): 1/1
              Auto negotiation: no
                         Speed: n/a
                       Address: 192.168.64.4 (DHCP4 via 192.168.64.1)
                       Gateway: 192.168.64.1
                           DNS: 192.168.64.1
               DHCP4 Client ID: 6:6a:7b:c5:3e:76

Mar 07 22:44:29 minikube systemd-networkd[180]: eth0: Link UP
Mar 07 22:44:29 minikube systemd-networkd[180]: eth0: Gained carrier
Mar 07 22:44:29 minikube systemd-networkd[180]: eth0: DHCPv4 address 192.168.64.4/24 via 192.168.64.1
```

If I run `resolvectl`, I get:

```
Global
           Protocols: +LLMNR +mDNS DNSOverTLS=opportunistic DNSSEC=no/unsupported
    resolv.conf mode: uplink
Fallback DNS Servers: 1.1.1.1 8.8.8.8 1.0.0.1 8.8.4.4 2606:4700:4700::1111 2001:4860:4860::8888
                      2606:4700:4700::1001 2001:4860:4860::8844

Link 2 (eth0)
    Current Scopes: DNS LLMNR/IPv4
         Protocols: +DefaultRoute +LLMNR -mDNS DNSOverTLS=opportunistic DNSSEC=no/unsupported
Current DNS Server: 192.168.64.1
       DNS Servers: 192.168.64.1

Link 3 (sit0)
Current Scopes: none
     Protocols: -DefaultRoute +LLMNR -mDNS DNSOverTLS=opportunistic DNSSEC=no/unsupported

Link 4 (docker0)
Current Scopes: none
     Protocols: -DefaultRoute +LLMNR -mDNS DNSOverTLS=opportunistic DNSSEC=no/unsupported
```

… so it still seems like we're using the DHCP DNS.

### Unexpected behaviour you saw

No matter what combinations of configuration I try, I cannot just configure "this is the DNS server to use" to `systemd`.

### Steps to reproduce the problem

(Described inline, above.)

### Additional Information

I've also tried,

```
Domains=~.
```
in `resolved.conf`, to no effect.

Further, I find the documentation confusing:

> [`Domains=`] define[s] a search path that preferably directs DNS queries to this interface.

… what is "this interface"? We're talking about `resolved`'s config, globally, but this reads as if we're in `/etc/systemd/network` & talking about a specific interface? (There are further uses of "this interface", again, without mention of _which_ interface. (N.b. that we're _not_ talking about `systemd-networkd`'s configuration, which if used like above, is per-interface.))

**Comment by roy-work** [comment]
This appears like it's probably a `minikube` bug.

I've been fully restarting the VM between configuration changes, on the assumption that you know, that ought to do it?

But a `systemctl restart systemd-resolved` causes changes to the outputs of the various `*ctl` commands, and to `/etc/resolv.conf`, and things look correct after that command.

That seems to indicate that `minikube` is transferring the files *after* the VM is already up & running, and thus, they're not taking effect.

I'm still perplexed on the the `Domains=` documentation language; that still looks like a bug to me. But, it seems the rest of this is not `systemd`'s fault.

**Comment by rpigott** [comment]
In addition to the per-link "scopes", sd-resolved has a supplementary global scope that is not associated with any interface. Requests to these servers can be performed in parallel with per-link unicast scopes, and are subject to the usual routing policy of the host. It can also have its own routing/search list, which matches queries to the global unicast scope in a manner similar to the per-link scopes. So, "to this interface" really means "to this scope", which is not the same only in this one case.

IMO most everyone is better off using only the per-link scopes or only the global unicast scope, possibly with per-link routing domains. Sending queries to multiple servers at once is not usually desirable behavior, so choosing a configuration such that no scopes overlap is the goal.

**Comment by rishta** [comment]
I think you were getting the DNS from DHCP. To aviod this, you need to have:
```
[DHCP]
UseDNS=false
```
in the `systemd-networkd` config.

**Comment by poettering** [comment]
Yeah, this seems to be a local misconfig. If you want to ignore the DHCP provided DNS config, then use UseDNS=false in the .network file for the network interface, to ignore it.
